File: src/main/java/com/codebetyars/skyhussars/engine/weapons/ProjectileManager.java
/*
 * Copyright (c) 2016, ZoltanTheHun
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * * Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.codebetyars.skyhussars.engine.weapons;

import com.codebetyars.skyhussars.engine.DataManager;
import com.jme3.scene.Node;
import java.util.ArrayList;
import java.util.List;

public class ProjectileManager {
    
    private List<Projectile> projectiles;
    private DataManager dataManager;
    private Node rootNode;
    
    public ProjectileManager(DataManager dataManager, Node node) {
        this.projectiles = new ArrayList<>();
        this.dataManager = dataManager;
        this.rootNode = node;
    }
    
    public void addProjectile(Bullet projectile) {
        rootNode.attachChild(dataManager.getBox().move(projectile.getLocation()));
        projectiles.add(projectile);
    }
    
    public void update(){
        for(Projectile projectile : projectiles){
            projectile.update();
        }
    }
}

File: src/main/scala/com/codebetyars/skyhussars/engine/weapons/ProjectileManager.scala
package com.codebetyars.skyhussars.engine.weapons

import com.codebetyars.skyhussars.engine.DataManager
import com.jme3.math.Vector3f
import com.jme3.scene.{Geometry, Node}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

import scala.collection.mutable

@Component
class ProjectileManager {

  @Autowired
  var rootNode: Node = _

  @Autowired
  var dataManager: DataManager = _

  var projectiles = mutable.Map[Projectile, Geometry]()


  def addProjectile(bullet: Bullet) {
    val geometry = dataManager.getBullet
    rootNode.attachChild(geometry)
    geometry.move(bullet.getLocation)

    projectiles += bullet -> geometry
  }

  def update(tpf: Float) {
    projectiles.foreach { case (projectile, geometry) =>
      projectile.update(tpf)
      geometry.setLocalTranslation(projectile.getLocation)
      geometry.lookAt(projectile.getVelocity.normalize(), Vector3f.UNIT_Y)
    }
  }

}
